File: extensions/shared/src/main/java/app/revanced/extension/youtube/patches/EnableDebuggingPatch.java
package app.revanced.extension.youtube.patches;

import java.util.Map;
import java.util.concurrent.ConcurrentHashMap;
import java.util.concurrent.ConcurrentMap;

import app.revanced.extension.shared.Logger;
import app.revanced.extension.shared.settings.BaseSettings;

@SuppressWarnings("unused")
public final class EnableDebuggingPatch {

    private static final ConcurrentMap<Long, Boolean> featureFlags
            = new ConcurrentHashMap<>(100, 0.75f, 1);

    public static boolean isFeatureFlagEnabled(long flag, boolean value) {
        if (value && BaseSettings.DEBUG.get()) {
            if (featureFlags.putIfAbsent(flag, true) == null) {
                Logger.printDebug(() -> "feature is enabled: " + flag);
            }
        }

        return value;
    }
}

File: patches/src/main/kotlin/app/revanced/patches/music/misc/debugging/EnableDebuggingPatch.kt
package app.revanced.patches.music.misc.debugging

import app.revanced.patches.music.misc.extension.sharedExtensionPatch
import app.revanced.patches.music.misc.settings.PreferenceScreen
import app.revanced.patches.music.misc.settings.settingsPatch
import app.revanced.patches.music.playservice.is_8_40_or_greater
import app.revanced.patches.music.playservice.is_8_41_or_greater
import app.revanced.patches.shared.misc.debugging.enableDebuggingPatch

@Suppress("unused")
val enableDebuggingPatch = enableDebuggingPatch(
    block = {
        dependsOn(
            sharedExtensionPatch,
            settingsPatch,
        )

        compatibleWith(
            "com.google.android.apps.youtube.music"(
                "7.29.52",
                "8.10.52",
                "8.37.56",
                "8.40.54",
            )
        )
    },
    // String feature flag does not appear to be present with YT Music.
    hookStringFeatureFlag = { false },
    // 8.40 has changes not worth supporting.
    hookLongFeatureFlag = { !is_8_40_or_greater || is_8_41_or_greater },
    hookDoubleFeatureFlag = { !is_8_40_or_greater || is_8_41_or_greater },
    preferenceScreen = PreferenceScreen.MISC,
)
